language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
             , vArg
             ]
 
--- | Infer the type of a 'Builtin'.
-inferTypeOfBuiltinM
+-- ###########################################
+-- ## Dealing with applications of builtins ##
+-- ###########################################
+
+-- | Infer the type components of a 'Builtin'. The annotation argument
+-- is required for the error if the name isn't found
+inferTypeComponentsOfBuiltinM
     :: (GShow uni, GEq uni, DefaultUni <: uni)
-    => Builtin ann -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+    => ann -> BuiltinName -> TypeCheckM uni ann (TypeComponents uni)
 -- We have a weird corner case here: the type of a 'BuiltinName' can contain 'TypedBuiltinDyn', i.e.
 -- a static built-in name is allowed to depend on a dynamic built-in type which are not required
 -- to be normalized. For dynamic built-in names we store a map from them to their *normalized types*,
 -- with the normalization happening in this module, but what should we do for static built-in names?
 -- Right now we just renormalize the type of a static built-in name each time we encounter that name.
-inferTypeOfBuiltinM (BuiltinName    _   name) = normalizeType $ typeOfBuiltinName name
+inferTypeComponentsOfBuiltinM ann (StaticBuiltinName name) = getStaticBuiltinTypeComponentsM ann name
 -- TODO: inline this definition once we have only dynamic built-in names.
-inferTypeOfBuiltinM (DynBuiltinName ann name) = lookupDynamicBuiltinNameM ann name
+inferTypeComponentsOfBuiltinM ann (DynBuiltinName name)    = lookupDynamicBuiltinTypeComponentsM ann name
+
+
+-- A mapping from type variables to types, recording type instantiations in an ApplyBuiltin term
+type TypeMapping uni = [(TyName, Normalized (Type TyName uni ()))]
+
+-- | Given a TypeMapping M and a type T, instantiate the type
+-- variables in T according to the assignemnts in M.
+substMapping
+    :: TypeMapping uni
+    -> Type TyName uni ()
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+substMapping mapping ty = do
+    let f ty1 (name, newty) = do
+          ty2 <- substNormalizeTypeM newty name ty1
+          return $ unNormalized ty2
+          -- ^ annoying: we have to normalise during subsitution, but
+          -- then we need a non-normalised type to perform subsequent
+          -- subsitutions on.
+    ty' <- foldM f (void ty) mapping
+    normalizeTypeM ty'
+
+-- | Work along the type parameters of a builtin application, building
+-- a mapping from type parameters to the types they're being
+-- instantiated at.
+checkBuiltinTypeArgs
+    :: ann
+    -> BuiltinName
+    -> [TyName]                -- type parameters
+    -> [Type TyName uni ann]   -- type instantations
+    -> TypeCheckM uni ann (TypeMapping uni)
+checkBuiltinTypeArgs = check [] where
+    check mapping ann bn (tyname:tynames) (tyarg:tyargs) = do
+       checkKindM  (typeAnn tyarg) tyarg (Type ()) -- NB: All of the type variables in TypeSchemes have kind *
+       tyarg' <- normalizeType (void tyarg)
```

Comment:
We need a note explaining how global uniqueness is preserved. Did you take any special care for handling global uniqueness? It looks correct to me, because you never wrap anything with `Normalized` explicitly and functions like `normalizeType` and `substNormalizeTypeM` ensure that global uniqueness is preserved.